### PR TITLE
Filtering all stream separate command

### PR DESCRIPTION
**Pull request by Yspadadden**
Is already in production for some time. Time to actually merge it with the other changes we made.

### CODE REVIEW

File: src/main/java/com/github/msemys/esjc/CatchUpSubscriptionSettings.java [hunk 1]
```diff
@@ -30,13 +30,17 @@
...
+
     public final Iterable<String> allowedEventTypes;
 
     private CatchUpSubscriptionSettings(Builder builder) {
         maxLiveQueueSize = builder.maxLiveQueueSize;
         resolveLinkTos = builder.resolveLinkTos;
         readBatchSize = builder.readBatchSize;
         allowedEventTypes = builder.allowedEventTypes;
+        maxSearchWindow = builder.maxSearchWindow;
+
```
Comment (daMupfel) [inline src/main/java/com/github/msemys/esjc/CatchUpSubscriptionSettings.java]: Remove line

File: src/main/java/com/github/msemys/esjc/CatchUpSubscriptionSettings.java [hunk 2]
```diff
@@ -30,13 +30,17 @@
      */
     public final int readBatchSize;
 
+    public final Integer maxSearchWindow;
```
Comment (daMupfel) [inline src/main/java/com/github/msemys/esjc/CatchUpSubscriptionSettings.java:37]: Maybe add documentation here, possibly also on allowedEventTypes

File: src/main/java/com/github/msemys/esjc/EventStore.java [hunk 3]
```diff
@@ -538,6 +531,47 @@ default Transaction continueTransaction(long transactionId) {
...
+     * @param maxCount       the maximum count of events to read, allowed range [1..4096].
+     * @param maxSearchWindow the maximum number of events the server should consider for filling a batch of maxCount events
+     * @param resolveLinkTos whether to resolve link events automatically.
+     * @return a {@code CompletableFuture} representing the result of this operation. The future's methods
+     * {@code get} and {@code join} can throw an exception with cause {@link CommandNotExpectedException},
+     * {@link NotAuthenticatedException}, {@link AccessDeniedException} or {@link ServerErrorException}
+     * on exceptional completion.
+     * @see #readAllEventsForwardFiltered(Position, int, int, boolean, Iterable, UserCredentials)
+     */
+    default CompletableFuture<AllEventsFilteredSlice> readAllEventsForwardFiltered(Position position,
```
Comment (daMupfel) [inline src/main/java/com/github/msemys/esjc/EventStore.java:548]: Missing javadoc for allowedEventTypes

File: src/main/java/com/github/msemys/esjc/EventStore.java [hunk 4]
```diff
@@ -538,6 +531,47 @@ default Transaction continueTransaction(long transactionId) {
...
+     * @param maxCount        the maximum count of events to read, allowed range [1..4096].
+     * @param maxSearchWindow the maximum number of events the server should consider for filling a batch of maxCount events
+     * @param resolveLinkTos  whether to resolve link events automatically.
+     * @param userCredentials user credentials to be used for this operation (use {@code null} for default user credentials).
+     * @return a {@code CompletableFuture} representing the result of this operation. The future's methods
+     * {@code get} and {@code join} can throw an exception with cause {@link CommandNotExpectedException},
+     * {@link NotAuthenticatedException}, {@link AccessDeniedException} or {@link ServerErrorException}
+     * on exceptional completion.
+     */
+    CompletableFuture<AllEventsFilteredSlice> readAllEventsForwardFiltered(Position position,
```
Comment (daMupfel) [inline src/main/java/com/github/msemys/esjc/EventStore.java:570]: Missing javadoc for allowedEventTypes

File: src/main/java/com/github/msemys/esjc/EventStoreTcp.java [hunk 5]
```diff
@@ -291,6 +290,22 @@ public Transaction continueTransaction(long transactionId, UserCredentials userC
...
+    @Override
+    public CompletableFuture<AllEventsFilteredSlice> readAllEventsForwardFiltered(
+        Position position,
+        int maxCount,
+        int maxSearchWindow,
+        boolean resolveLinkTos,
+        Iterable<String> allowedEventTypes,
+        UserCredentials userCredentials) {
+
+        checkArgument(BATCH_SIZE_RANGE.contains(maxCount), "maxCount is out of range. Allowed range: %s.", BATCH_SIZE_RANGE.toString());
```
Comment (daMupfel) [inline src/main/java/com/github/msemys/esjc/EventStoreTcp.java:302]: Shouldn't there be a check for maxSearchWindow as well?
maxSearchWindow >= maxCount

**Review by daMupfel — CHANGES_REQUESTED** [review]

File: src/main/java/com/github/msemys/esjc/EventStoreTcp.java (hunk 5 above)
Comment (Yspadadden) [reply src/main/java/com/github/msemys/esjc/EventStoreTcp.java:302]: I don't think this check is necessary. The server does not enforce this contraint. If you want 50 events and set a window of 5 you will get  5 events or less. Not very sensible, but also not illegal.

**Review by daMupfel — APPROVED** [review]

**Pull request merged**